### ISSUE TITLE
Adding DESC to data integration benchmarking

### ISSUE THREAD

**Issue by LuckyMD** [eleozzr/desc#28]
Hi @eleozzr,

We were thinking about adding DESC to our benchmark of data integration tools (https://github.com/theislab/scib). We would be running our own pre-processing for the input to DESC for this, which is reliant on Scanpy version 1.4.5+. Do you think it would be possible to use just the `desc.train()` function if we remove the Scanpy requirement and install via github? Would this also be okay for using Keras 2.2.4?

Also, to compare the methods properly we would not be able to use the clustering output you provide, but instead we would use the embedding at a default clustering resolution (resolution=0.8 as in your tutorial). Would this be a suitable way of evaluating DESC?

Kind regards,

**Comment by eleozzr** [comment]
Because DESC's dependency on other packages, such as TensorFlow,scanpy and keras, we are updating and testing our desc algorithm to be compatible with  tensorflow2.0, scanpy 1.4.5+. Hopefully, the latest algorithm can be uploaded into GitHub and PyPI tomorrow.  

The single resolution (0.8 or 1.0) is ok for DESC.

Thanks.

**Comment by LuckyMD** [comment]
Hi @eleozzr,

That is great news! If you have it online by tomorrow, we will be able to add it to the benchmark on time :). I am looking forward to a `setup.py` with keras>2.1 and Scanpy>1.3.6 as dependencies :). I hope you will still be able to use tensorflow 1.X as well though, and that it's not exclusive to tensorflow 2.

**Comment by eleozzr** [comment]
Hi LuckyMD, 

I have already updated our desc algorithm. 

1. For  `tensorflow 1*`, we released `desc(2.0.3)`. Please see our jupyter notebook example [desc_2.0.3_paul.ipynb](https://github.com/eleozzr/desc/blob/master/desc_2.0.3_paul.ipynb)
2. For  `tensorflow 2*`, we released `desc(2.1.1)`. Please see our jupyter notebook example [desc_2.1.1_paul.ipynb](https://github.com/eleozzr/desc/blob/master/desc_2.1.1_paul.ipynb)

Hope this helps. Thanks.

**Comment by LuckyMD** [comment]
hi @eleozzr,

Thanks for the updates. However, I see that the Scanpy version is still capped to 1.4.4 in version 2.0.3. Is it possible to remove this?

**Comment by LuckyMD** [comment]
Also, what is the limit to python 3.7 compatability?

**Comment by eleozzr** [comment]
> Also, what is the limit to python 3.7 compatability?

All the script was tested in python3.5 or python3.6, but I think the script works in python3.7.

**Comment by LuckyMD** [comment]
Are you still planning to fix the install requirements in: https://github.com/eleozzr/desc/blob/9acb04702227d684d7e81a338320c59f7a423459/setup.py#L21-L23

Then I could just install from `pip` directly rather than making a git fork and installing via that.

**Comment by eleozzr** [comment]
> Are you still planning to fix the install requirements in:
> 
> https://github.com/eleozzr/desc/blob/9acb04702227d684d7e81a338320c59f7a423459/setup.py#L21-L23
> 
> Then I could just install from `pip` directly rather than making a git fork and installing via that.

Could you install by `pip install desc==2.1.1 or 2.0.3`? 
I haven't figured out why the code in Github is not the latest.

**Comment by LuckyMD** [comment]
I hadn't tried to install via `pip` yet, as the `setup.py` with the 2.0.3 tag still showed these dependencies. I will give it a go and report back.

**Comment by LuckyMD** [comment]
Install for 2.0.3 worked, thanks :). Is there a reason you are limited to keras 2.1? Just a question... I can work with that as well ;).

**Comment by LuckyMD** [comment]
Hi @eleozzr,

I have another question about using DESC. I can't see where you pass batch information to the algorithm so that it can perform data integration. Do you explicitly integrate data across batches or just produce a low-dimensional embedding that is less affected by batch?

If DESC doesn't explicitly do data integration, but only produces a low-dimensional embedding which is less affected by batch effects than the high-dimensional data, maybe we shouldn't be comparing it to other data integration tools? I guess that comparison might not be fair to DESC.

What do you think?

**Comment by eleozzr** [comment]
> with



> Install for 2.0.3 worked, thanks :). Is there a reason you are limited to keras 2.1? Just a question... I can work with that as well ;).

Sometimes, the version of TensorFlow and Keras needs to match.  So I directly limit the version to avoid unnecessary issues due to the unmatch of Tensorflow and Keras.

**Comment by eleozzr** [comment]
> Hi @eleozzr,
> 
> I have another question about using DESC. I can't see where you pass batch information to the algorithm so that it can perform data integration. Do you explicitly integrate data across batches or just produce a low-dimensional embedding that is less affected by batch?
> 
> If DESC doesn't explicitly do data integration, but only produces a low-dimensional embedding which is less affected by batch effects than the high-dimensional data, maybe we shouldn't be comparing it to other data integration tools? I guess that comparison might not be fair to DESC.
> 
> What do you think?

The only information for batch information for desc is the batch id. Generally speaking, you should scale data within each batch instead of scaling across all cells.

**Comment by LuckyMD** [comment]
Okay, then I will add batch-specific scaling and then make a PR for the Benchmarking data integration repo [here](www.github.com/theislab/scib). Would it be okay if I tagged you in that PR so you could check that DESC is used as you think is correct?

**Comment by eleozzr** [comment]
> Thanks for the example! The parametrization is slightly different to your example notebook. I will use something closer to this if you think this is a better default parametrization for datasets of 10k+ cells?
> 
> I will then add batch-specific scaling and then make a PR for the Benchmarking data integration repo [here](www.github.com/theislab/scib). Would it be okay if I tagged you in that PR so you could check that DESC is used as you think is correct?

If you only need the embedding of desc, you can set `do_tsne=False` and `do_umap=False`. 

And it will be okay if you tagged me.

**Comment by LuckyMD** [comment]
> If you only need the embedding of desc, you can set `do_tsne=False` and `do_umap=False`.

Yes, I've already changed this :). I will test the code, make a PR and then you can tell me if I'm doing something stupid ;).

**Comment by LuckyMD** [comment]
Thing seem to be running for me now, thanks! I just quickly wanted to highlight 2 things:

1. Installing DESC also requires `pydot` and `GraphViz` (not the python package, but the C binaries), which are not automatically installed with the keras dependency (this is a keras issue, but maybe worth highlighting somewhere).
2. Installing GraphViz via `conda` installed `keras` version 2.3.1 for me, but the code still ran through (I am still using tensorflow version 1.14). I assume that you don't really need the pinned version you have in `setup.py`.

**Comment by eleozzr** [comment]
> 2\. setup

Thanks.

**Comment by LuckyMD** [comment]
It would be great if you could look over our PR here: https://github.com/theislab/scib/pull/131

Thanks!

**Comment by LuckyMD** [comment]
Hi @eleozzr,
We are still having an issue with saving the network weights. I have turned off saving weights via `save_encoder_weights=False`, but then I set `save_dir=tmp_dir` as you suggest in the defaults. However, we have no permissions to save on a local directory in our server. Is there a way to turn off saving any files?

**Comment by LuckyMD** [comment]
It would be great to get an input on the above question of how to turn off saving the weights.
